store/smt/store.go

```diff
@@ -0,0 +1,157 @@
+package smt
+
+import (
+	"crypto/sha256"
+	"io"
+	"sync"
+
+	"github.com/cosmos/cosmos-sdk/store/cachekv"
+	"github.com/cosmos/cosmos-sdk/store/tracekv"
+	"github.com/cosmos/cosmos-sdk/store/types"
+	abci "github.com/tendermint/tendermint/abci/types"
+	dbm "github.com/tendermint/tm-db"
+
+	"github.com/lazyledger/smt"
+)
+
+var (
+	_ types.KVStore                 = (*Store)(nil)
+	_ types.CommitStore             = (*Store)(nil)
+	_ types.CommitKVStore           = (*Store)(nil)
+	_ types.Queryable               = (*Store)(nil)
+	_ types.StoreWithInitialVersion = (*Store)(nil)
+)
+
+// Store Implements types.KVStore and CommitKVStore.
+type Store struct {
+	tree *smt.SparseMerkleTree
+	db   dbm.DB
+
+	mtx sync.RWMutex
+}
+
+func NewStore(underlyingDB dbm.DB) *Store {
+	return &Store{
+		tree: smt.NewSparseMerkleTree(underlyingDB, sha256.New()),
+		db:   underlyingDB,
+	}
+}
+
+// KVStore interface below:
+
+func (s *Store) GetStoreType() types.StoreType {
+	return types.StoreTypeSMT
+}
+
+// CacheWrap branches a store.
+func (s *Store) CacheWrap() types.CacheWrap {
+	return cachekv.NewStore(s)
+}
+
+// CacheWrapWithTrace branches a store with tracing enabled.
+func (s *Store) CacheWrapWithTrace(w io.Writer, tc types.TraceContext) types.CacheWrap {
+	return cachekv.NewStore(tracekv.NewStore(s, w, tc))
+}
+
+// Get returns nil iff key doesn't exist. Panics on nil key.
+func (s *Store) Get(key []byte) []byte {
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+	val, err := s.tree.Get(key)
+	// TODO(tzdybal): how to handle this err?
+	if err != nil {
+		return nil
```

Comment:
SMT propagates errors from underlying storage, so we should panic (and probably log before)